File: src/main/java/at/yawk/fiction/impl/fimfiction/StoryParser.java
package at.yawk.fiction.impl.fimfiction;

import at.yawk.fiction.HtmlText;
import at.yawk.fiction.Image;
import at.yawk.fiction.impl.PageParser;
import java.net.URI;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;
import org.jsoup.nodes.Element;
import org.jsoup.nodes.Node;
import org.jsoup.select.Elements;

/**
 * @author yawkat
 */
class StoryParser extends PageParser<FimStory> {
    @Override
    protected FimStory create() {
        return new FimStory();
    }

    @Override
    protected void parse(Element root, FimStory target) throws Exception {
        // only applicable outside a search when the root is a html element
        Element storyContainer = root.select(".story_container").first();

        target.setId(Integer.parseInt(root.select(".story_container").attr("data-story")));
        Element storyName = root.select(".story_name").first();
        target.setTitle(storyName.text());
        target.setUri(URI.create(storyName.absUrl("href")));

        FimAuthor author = new FimAuthor();
        Element authorLink = root.select(".author a").first();
        String authorHref = authorLink.attr("href");
        author.setId(extractGroup(authorHref, "/user/(\\d)+/.*"));
        author.setName(authorLink.text());
        target.setAuthor(author);

        Element descriptionElement = storyContainer.select(".description").first();
        Element storyImage = descriptionElement.select(".story_container__story_image img").first();
        if (storyImage != null) {
            Image image = new Image();
            if (storyImage.hasAttr("data-src")) {
                image.setImageUrl(URI.create(storyImage.absUrl("data-src")));
            }
            if (storyImage.hasAttr("src")) {
                image.setThumbnailUrl(URI.create(storyImage.absUrl("src")));
            }
            target.setImage(image);
        }

        Element descriptionText = descriptionElement.select("> .description-text").first();
        if (descriptionText != null) {
            HtmlText html = new HtmlText();
            html.setHtml(descriptionText.html());
            target.setDescription(html);
        }

        List<FimTag> tags = new ArrayList<>();

        for (Element tagElement : storyContainer.select(".story-tags a")) {
            FimTag tag = new FimTag();
            tag.setId(tagElement.attr("data-tag"));
            tag.setName(tagElement.text());
            tag.setIcon(null);
            tags.add(tag);
        }

        target.setTags(tags);

        Elements footer = storyContainer.select(".chapters-footer");
        if (footer.select(".completed-status-incomplete") != null) {
            target.setStatus(FimStatus.INCOMPLETE);
        } else if (footer.select(".completed-status-complete") != null) {
            target.setStatus(FimStatus.COMPLETED);
        } else if (footer.select(".completed-status-hiatus") != null) {
            target.setStatus(FimStatus.ON_HIATUS);
        } else if (footer.select(".completed-status-cancelled") != null) {
            target.setStatus(FimStatus.CANCELLED);
        }

        List<FimChapter> chapters = new ArrayList<>();
        for (Element chapterContainer : storyContainer.select(".chapters > li > div")) {
            if (chapterContainer.hasClass("chapter_expander")) { continue; }

            Element chapterLink = chapterContainer.select(".chapter-title").first();

            FimChapter chapter = new FimChapter();
            chapter.setId(parseIntLenient(extractGroup(
                    chapterContainer.select(".word_count .drop-down a").first().attr("href"), "/chapters/download/(\\d+)/txt")));
            chapter.setIndex(parseIntLenient(extractGroup(
                    chapterLink.attr("href"), ".*/story/\\d+/[^/]*/(\\d+)(:?/.*)?")) - 1);
            chapter.setName(chapterLink.text());
            chapter.setWordCount(parseIntLenient(chapterContainer.select(".word_count .word-count-number").first().text()));

            Element chapterReadIcon = chapterContainer.select(".chapter-read-icon").first();
            if (chapterReadIcon != null) {
                chapter.setRead(chapterReadIcon.hasClass("chapter-read"));
            } else {
                chapter.setRead(null);
            }

            chapters.add(chapter);
        }
        target.setChapters(chapters);

        // todo: parse first published + last modified
    }
}

File: src/main/kotlin/at/yawk/fiction/impl/fimfiction/StoryParser.kt
package at.yawk.fiction.impl.fimfiction

import at.yawk.fiction.*
import at.yawk.fiction.impl.PageParser
import org.jsoup.nodes.Element
import java.net.URI
import java.net.URL
import java.util.*

/**
 * @author yawkat
 */
internal object StoryParser : PageParser<Story>() {
    override fun parse(root: Element): Story {
        // only applicable outside a search when the root is a html element
        val storyContainer = root.select(".story_container").first()

        val contentBox = storyContainer.select(".story_content_box")
        val id = parseIntLenient(contentBox.attr("id"))
        val storyName = contentBox.select(".story_name").first()
        val title = storyName.text()
        val uri = URI.create(storyName.absUrl("href"))

        val authorLink = contentBox.select(".author a")
        val authorHref = authorLink.attr("href")
        val author = Author(
                name = authorLink.text(),
                providerData = FimAuthorProviderData(id = authorHref.substring(authorHref.lastIndexOf('/') + 1))
        )

        val descriptionElement = storyContainer.select(".description").first()
        val storyImage = descriptionElement.select(".story_image img").first()
        var image: Image? = null
        if (storyImage != null) {
            var imageUri: URI? = null
            if (storyImage.hasAttr("data-src")) {
                imageUri = URI.create(storyImage.absUrl("data-src"))
            }
            var thumbnailUri: URI? = null
            if (storyImage.hasAttr("src")) {
                thumbnailUri = URI.create(storyImage.absUrl("src"))
            }
            if (imageUri != null || thumbnailUri != null) {
                image = Image(
                        imageUrl = imageUri ?: thumbnailUri,
                        thumbnailUrl = thumbnailUri
                )
            }
        }

        var description: HtmlText? = null

        val separator = descriptionElement.select(">hr").first()
        if (separator != null) {
            val descriptionHtml = StringBuilder()
            var e = separator.nextSibling()
            while (e != null) {
                descriptionHtml.append(e.outerHtml())
                e = e.nextSibling()
            }

            description = HtmlText(descriptionHtml.toString())
        }

        val tags = ArrayList<FimTag>()

        for (categoryElement in descriptionElement.select("> .story_category")) {
            tags.add(FimTag(
                    id = extractGroup(categoryElement.attr("href"), ".*tags\\[\\]=(.+)")!!,
                    name = categoryElement.text()
            ))
        }

        for (characterIcon in root.select(".extra_story_data .inner_data").first().select("> .character_icon")) {
            tags.add(FimTag(
                    // we can't find out id here :(
                    name = characterIcon.attr("title"),
                    icon = URL(characterIcon.select("img").first().absUrl("src"))
            ))
        }

        val statusElement = storyContainer.select(".chapters .bottom > span").first()
        val status = when (statusElement.attr("title")) {
            "Complete" -> FimStatus.COMPLETED
            "Incomplete" -> FimStatus.INCOMPLETE
            "On Hiatus" -> FimStatus.ON_HIATUS
            "Cancelled" -> FimStatus.CANCELLED
            else -> null
        }

        val chapters = ArrayList<Chapter>()
        for (chapterContainer in storyContainer.select(".chapters .chapter_container")) {
            if (chapterContainer.hasClass("chapter_expander")) {
                continue
            }

            val chapterLink = chapterContainer.select(".chapter_link").first()

            val id = parseIntLenient(extractGroup(
                    chapterContainer.select(".download_container a").first().attr("href"), ".*chapter=(\\d+)"))
            val index = parseIntLenient(extractGroup(
                    chapterLink.attr("href"), ".*/story/\\d+/(\\d+)(:?/.*)?")) - 1
            val name = chapterLink.text()
            val wordCount = parseIntLenient(chapterContainer.select(".word_count").first().ownText())

            val chapterReadIcon = chapterContainer.select(".chapter-read-icon").first()
            val read = chapterReadIcon?.hasClass("chapter-read")

            chapters.add(Chapter(
                    name = name,
                    read = read,
                    providerData = FimChapterProviderData(
                            id = id,
                            index = index,
                            wordCount = wordCount
                    )
            ))
        }

        // todo: parse first published + last modified

        return Story(
                title = title,
                author = author,
                chapters = chapters,
                description = description,
                image = image,
                uri = uri,
                providerData = FimStoryProviderData(
                        id = id,
                        status = status,
                        tags = tags
                )
        )
    }
}
